From fed140a53170518f5adda3e1363c441aa9f66a07 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 14:58:27 -0400
Subject: [infiniband] usnic: Fix smatch endianness error

Message-id: <e791c5c4bb0103e31a9e1b34f9141532b939503c.1411569661.git.dledford@redhat.com>
Patchwork-id: 96588
O-Subject: [Patch RHEL7 25/29] IB/usnic: Fix smatch endianness error
Bugzilla: 916382
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Upinder Malhi <umalhi@cisco.com>

Error reported at http://marc.info/?l=linux-rdma&m=138995755801039&w=2

Fix short to int cast for big endian systems.

Signed-off-by: Upinder Malhi <umalhi@cisco.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit f809309a251a13bd97cc189c3fa428782aab9716)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/usnic/usnic_ib_qp_grp.c b/drivers/infiniband/hw/usnic/usnic_ib_qp_grp.c
index 7ecc606..f8dfd76 100644
--- a/drivers/infiniband/hw/usnic/usnic_ib_qp_grp.c
+++ b/drivers/infiniband/hw/usnic/usnic_ib_qp_grp.c
@@ -629,6 +629,7 @@ static int qp_grp_id_from_flow(struct usnic_ib_qp_grp_flow *qp_flow,
 {
  enum usnic_transport_type trans_type = qp_flow->trans_type;
  int err;
+ uint16_t port_num = 0;
 
  switch (trans_type) {
  case USNIC_TRANSPORT_ROCE_CUSTOM:
@@ -637,9 +638,15 @@ static int qp_grp_id_from_flow(struct usnic_ib_qp_grp_flow *qp_flow,
  case USNIC_TRANSPORT_IPV4_UDP:
   err = usnic_transport_sock_get_addr(qp_flow->udp.sock,
        NULL, NULL,
-       (uint16_t *) id);
+       &port_num);
   if (err)
    return err;
+  /*
+   * Copy port_num to stack first and then to *id,
+   * so that the short to int cast works for little
+   * and big endian systems.
+   */
+  *id = port_num;
   break;
  default:
   usnic_err("Unsupported transport %u\n", trans_type);
-- 
1.7.1